Sprint 3/step 1

anufrp opened the pull request.

anufrp committed:
Роутинг - начало

Подключение либы, первые роуты в APP

anufrp committed:
Страница логина - начало

Верстка страницы логина

anufrp committed:
Почти все страницы + начало роутинга

Без страницы 404 и детали ингредиента

anufrp committed:
Почти весь роутинг и общение с API

anufrp committed:
Весь роутинг и защищенные маршруты

anufrp committed:
Отдельная страница ингредиента

anufrp committed:
Модалка\отдельная страница ингредиента

anufrp committed:
История заказа в профиле + роутинг там же

anufrp committed:
fix активной иконки профиля

gennady-bars commented on src/utils/create-order.js:
```diff
@@ -1,13 +1,15 @@
 import request from "./make-request";
 
 export function createOrder(url, ingredients) {
     let options = {
```
Если переменная объявлена через `let`, её значение должно быть напрямую перезаписано.
Если переменная **не перезаписывает** своё значение, она объявляется через `const`.


Все переменные нужно объявлять по возможности через `const`.
В больших функциях бывает так, что Вы можете не заметить, если переменная переназначится, и это может оказаться причиной багов. Особенно в замыканиях. 
`const` дает Вам уверенность в том, что Вы всегда будете видеть одно и то же значение.

gennady-bars commented on src/services/actions/user.js:
```diff
@@ -0,0 +1,265 @@
+            else { 
+                dispatch({type: REGISTER_USER_FAILED});
+            }
+        });
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

**Нужно исправить это везде по коду в конце запросов.**

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,119 @@
+        }));
+
+    const [email, setEmail] = useState(null);
+    const [password, setPassword] = useState(null);
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,119 @@
+    if (user.name) {
+        return (
+          <Redirect
+            to={ location.state?.from || '/' }
```
Хорошая переадресация

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,119 @@
+  return (
+    <div className={styles.wrapper}>
+        <div className={styles.content}>
+        <form className={`${styles.form} pb-20`} onSubmit={formSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -42,7 +42,7 @@ export default function BurgerIngredients() {
 
     useEffect(() => {
-        dispatch(getIngredients());
+        if(ingredients.length === 0) dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -1,13 +1,33 @@
+    }
+
+    useEffect(() => {
+        if(ingredients.length === 0) dispatch(getIngredients()); //получить ингредиенты, если страница была перезагружена
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/ingredients/ingredients.jsx:
```diff
@@ -0,0 +1,58 @@
+
+    const closeModal = () => {
+        // dispatch({type: DROP_CHEK_EMAIL_ERROR});
+        dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/ingredients/ingredients.jsx:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    useEffect(() => {
+        if(ingredients.length === 0) dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/protected-route.jsx:
```diff
@@ -0,0 +1,57 @@
+            <Redirect
+                to={{
+                pathname: '/login',
+                state: { from: location }
```
Хорошо сделан `ProtectedRoute`

gennady-bars commented on src/components/protected-route.jsx:
```diff
@@ -0,0 +1,57 @@
+//         />
+//       );
+//     }
+
```
**Можно лучше**

Можно сделать `ProtectedRoute` примерно так:

```jsx
const ProtectedRoute = ({ onlyForAuth, children, ...rest }) => {
  const isAuthorized = getCookie("accessToken");
  const location = useLocation();

  if (!onlyForAuth && isAuthorized) {
    const { from } = location.state || { from: { pathname: "/" } };
    return (
      <Route {...rest}>
        <Redirect to={from} />
      </Route>
    );
  }

  if (onlyForAuth && !isAuthorized) {
    return (
      <Route {...rest}>
        <Redirect to={{ pathname: "/login", state: { from: location } }} />
      </Route>
    );
  }

  return <Route {...rest}>{children}</Route>;
};

```

Этот рут может и переадресовывать пользователя туда, откуда он пришел, если он уже авторизован

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -8,21 +10,51 @@ import styles from './app.module.css';
+
+    const location = useLocation();
+    const history = useHistory();
+    const background = location.state && location.state.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Хорошо сделан `ProtectedRoute`
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

но есть некоторые недочеты: 

* в попапе ингредиента должна быть надпись https://disk.yandex.ru/i/9migBS7UF8n4hA 
* Если переменная **не перезаписывает** своё значение, она объявляется через `const`.
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* `dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.





**Можно лучше**


*  при клике на **логотип** обычно идет переход на главную страницу


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

anufrp committed:
Заголовок модалки Детали ингредиентов

anufrp committed:
Перенос получения ингредиентов в компонент App

anufrp committed:
catch для всех запросов к API

+кликабельный логотип

anufrp committed:
Кастомный хук для полей форм

useForm для обработки полей форм на всех страницах с формами

anufrp committed:
Рефакторинг ProtectedRoute

В ProtecterRoute теперь работает защита от авторизованного юзера. + финальная чистка кода

anufrp commented:
Исправил все недочеты + улучшил в указанных местах

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* когда перезагружаю страницу с открытым попапом и потом его закрываю, то весь контент съезжает в бок и появляется горизонтальный скролл https://disk.yandex.ru/i/apRJp0AjDjg6Qw

anufrp committed:
Ширина Body при закрытии модалки после перезагрузки

убрал хардкод ширины body

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
